File: problems/0015/solution.js
const { getPrimeFactors } = require("../../helpers/functions.js")

module.exports = (size) => {
    if (!Number.isInteger(size)) return 0

    let numeratorDict = {}
    let denominatorDict = {}

    for (let i = 0; i < size; i++) {
        const numerator = i + size + 1
        const denominator = i + 1

        const numberatorBases = getPrimeFactors(numerator)
        const denominatorBases = getPrimeFactors(denominator)

        numberatorBases.map((it) => {
            const maxExponent = Math.floor(Math.log(numerator) / Math.log(it))

            for (let e = maxExponent; e >= 1; e--) {
                const p = Math.pow(it, e)
                if (numerator % p === 0) {
                    if (numeratorDict[it]) {
                        numeratorDict[it] += e
                    } else {
                        numeratorDict[it] = e
                    }
                    break
                }
            }
        })

        denominatorBases.map((it) => {
            const maxExponent = Math.floor(Math.log(denominator) / Math.log(it))

            for (let e = maxExponent; e >= 1; e--) {
                const p = Math.pow(it, e)
                if (denominator % p === 0) {
                    if (denominatorDict[it]) {
                        denominatorDict[it] += e
                    } else {
                        denominatorDict[it] = e
                    }
                    break
                }
            }
        })
    }

    const newNumeratorDict = {}
    const newDenominatorDict = {}

    Object.keys(numeratorDict).map((base) => {
        if (denominatorDict[base]) {
            newNumeratorDict[base] = numeratorDict[base] - denominatorDict[base]
        } else {
            newNumeratorDict[base] = numeratorDict[base]
        }
    })

    Object.keys(denominatorDict).map((base) => {
        if (numeratorDict[base]) {
            newDenominatorDict[base] = 0
        } else {
            newDenominatorDict[base] = denominatorDict[base]
        }
    })

    let numerator = 1
    let denominator = 1

    Object.keys(newNumeratorDict).map((base) => {
        numerator *= Math.pow(base, newNumeratorDict[base])
    })

    Object.keys(newDenominatorDict).map((base) => {
        denominator *= Math.pow(base, newDenominatorDict[base])
    })

    return numerator / denominator
}

File: problems/0015/solution.cpp
#include <iostream>
#include <vector>
#include <map>
#include "input.h"
#include "../../helpers/functions.cpp"
using namespace std;

int main()
{
    map<int, int> numeratorDict;
    map<int, int> denominatorDict;

    for (int i = 0; i < SIZE; i++)
    {
        int numerator = i + SIZE + 1;
        int denominator = i + 1;

        vector<int> numeratorBases = getPrimeFactors(numerator);
        vector<int> denominatorBases = getPrimeFactors(denominator);

        for (vector<int>::iterator it = numeratorBases.begin(); it != numeratorBases.end(); ++it)
        {
            int maxExponent = floor(log(numerator) / log(*it));

            for (int e = maxExponent; e >= 1; e--)
            {
                int p = pow(*it, e);
                if (numerator % p == 0)
                {
                    if (numeratorDict[*it] != 0)
                    {
                        numeratorDict[*it] = numeratorDict[*it] + e;
                    }
                    else
                    {
                        numeratorDict[*it] = e;
                    }
                    break;
                }
            }
        }

        for (vector<int>::iterator it = denominatorBases.begin(); it != denominatorBases.end(); ++it)
        {
            int maxExponent = floor(log(denominator) / log(*it));

            for (int e = maxExponent; e >= 1; e--)
            {
                int p = pow(*it, e);
                if (denominator % p == 0)
                {
                    if (denominatorDict[*it] != 0)
                    {
                        denominatorDict[*it] = denominatorDict[*it] + e;
                    }
                    else
                    {
                        denominatorDict[*it] = e;
                    }
                    break;
                }
            }
        }
    }

    map<int, int> newNumeratorDict;
    map<int, int> newDenominatorDict;

    for (map<int, int>::iterator itr = numeratorDict.begin(); itr != numeratorDict.end(); ++itr)
    {
        int base = itr->first;
        if (denominatorDict[base] != 0)
        {
            newNumeratorDict[base] = numeratorDict[base] - denominatorDict[base];
        }
        else
        {
            newNumeratorDict[base] = numeratorDict[base];
        }
    }

    for (map<int, int>::iterator itr = denominatorDict.begin(); itr != denominatorDict.end(); ++itr)
    {
        int base = itr->first;
        if (numeratorDict[base] != 0)
        {
            newDenominatorDict[base] = 0;
        }
        else
        {
            newDenominatorDict[base] = denominatorDict[base];
        }
    }

    long long numerator = 1;
    long long denominator = 1;

    for (map<int, int>::iterator itr = newNumeratorDict.begin(); itr != newNumeratorDict.end(); ++itr)
    {
        numerator *= pow(itr->first, itr->second);
    }

    for (map<int, int>::iterator itr = newDenominatorDict.begin(); itr != newDenominatorDict.end(); ++itr)
    {
        denominator *= pow(itr->first, itr->second);
    }

    cout << "C++: " << numerator / denominator << "  - (solution)" << endl;

    return 0;
}
